extras/upgrade_fw.py

```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/python3
+import subprocess
+import argparse
+import logging
+from dmi import Dmi
+from devices import *
+
+SSH_OPTS = "-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null"
+target_device_username = "ubuntu"
+
+
+def all_subclasses(cls):
+    return cls.__subclasses__() + [
+        g for s in cls.__subclasses__() for g in all_subclasses(s)
+    ]
+
+
+def detect_device(ip: str, user: str, password: str = "", **options):
+    """
+    detect device's firmware upgrade type by checking on DMI data
+
+    :returns: device object
+    :rtype: object of subclass of AbstractDevice
+    """
+    temp_device = LVFSDevice(ip, user, password)
+    run_ssh = temp_device.run_cmd
+    devices = all_subclasses(AbstractDevice)
+    try:
+        dmi_chassis_vendor = "sudo cat /sys/class/dmi/id/chassis_vendor"
+        dmi_chassis_type = "sudo cat /sys/class/dmi/id/chassis_type"
+        rc1, vendor_string, stderr1 = run_ssh(
+            dmi_chassis_vendor, raise_stderr=False
+        )
+        rc2, type_string, stderr2 = run_ssh(
+            dmi_chassis_type, raise_stderr=False
+        )
+
+        err_msg = ""
+        if rc1 != 0:
+            err_msg = vendor_string + stderr1 + "\n"
+        if rc2 != 0:
+            err_msg = err_msg + type_string + stderr2
+        if err_msg:
+            err_msg = (
+                "Unable to detect device vendor/type due to lacking of dmi info.\n"
+                + err_msg
+            )
+            logger.error(err_msg)
+            print(err_msg)
+            exit(1)
+
+        type_index = int(type_string)
+        upgrade_type = Dmi.chassis_types[type_index]
+        try:
+            dev = [
+                dev
+                for dev in devices
+                if dev.fw_update_type in upgrade_type
+                and any(x == vendor_string for x in dev.vendor)
+            ][0]
+            logger.info(f"{ip} is a {vendor_string} {dev.__name__}")
+        except IndexError:
+            err_msg = f"{vendor_string} {Dmi.chassis_names[type_index]} Device is not in current support scope"
+            logger.error(err_msg)
+            print(err_msg)
+            exit(1)
+
+        if issubclass(dev, LVFSDevice):
+            return dev(ip, user, password)
+        elif issubclass(dev, OEMDevice):
+            if not (
+                "bmc_ip" in options
+                and "bmc_user" in options
+                and "bmc_password" in options
+            ):
+                print(
+                    "Please provide $BMC_IP, $BMC_USER, $BMC_PASSWORD for this device"
+                )
+                exit(1)
+            return dev(
+                options.get("bmc_ip"),
+                options.get("bmc_user"),
+                options.get("bmc_password"),
+            )
+    except subprocess.CalledProcessError as e:
+        logger.error(e.output)
+        raise RuntimeError(e.output)
```

Comment:
```suggestion
        raise RuntimeError(e.output) from e
```

From [Python docs](https://docs.python.org/3/tutorial/errors.html#exception-chaining): 
> To indicate that an exception is a direct consequence of another, the raise statement allows an optional from clause

But you can keep it as is if there's no direct consequence

Reply:
Keep it as is since there's no direct consequence.